src/components/utils/src/threads/thread_posix.cc

```diff
@@ -258,48 +260,57 @@ void Thread::yield() {
 void Thread::stop() {
   SDL_AUTO_TRACE();
   sync_primitives::AutoLock auto_lock(state_lock_);
-
-  stopped_ = true;
-
-  SDL_DEBUG("Stopping thread #" << handle_ << " \"" << name_ << " \"");
-
-  if (delegate_ && isThreadRunning_) {
-    delegate_->exitThreadMain();
-  }
-
-  SDL_DEBUG("Stopped thread #" << handle_ << " \"" << name_ << " \"");
+  StopUnsafe();
 }
 
 void Thread::join(const bool force_stop) {
   SDL_AUTO_TRACE();
   DCHECK_OR_RETURN_VOID(!IsCurrentThread());
-
-  if (force_stop) {
-    stop();
-  }
-
   sync_primitives::AutoLock auto_lock(state_lock_);
-
-  if (!force_stop) {  // then wait on going iteration to finish
-    while (do_iteration_ && thread_created_) {
+  // if thread still exist at all
+  if (kThreadStateNone != thread_state_) {
+    if (force_stop) {
+      StopUnsafe();
+    }
+    // force thread to execute last command if there is a pending one
+    if ((kThreadStateIdle == thread_state_) &&
+        (kThreadCommandNone != thread_command_)) {
+      run_cond_.NotifyOne();
+    }
+    while ((kThreadStateRunning == thread_state_) ||
+           (kThreadCommandNone != thread_command_)) {
       state_cond_.Wait(auto_lock);
     }
   }
+}
 
-  stopped_ = true;
-  do_iteration_ = true;
-  run_cond_.NotifyOne();
+void Thread::StopUnsafe() {
+  SDL_AUTO_TRACE();
+  thread_command_ = kThreadCommandStop;
```

Comment:
@anosach-luxoft, fixed in [ce7130e](https://github.com/smartdevicelink/sdl_core/pull/707/commits/ce7130e0fa97a34d5fb143ff0764557cee82eb68)